rules/internal/utils/declarative_friendly.go

```diff
@@ -22,155 +22,112 @@ import (
 	apb "google.golang.org/genproto/googleapis/api/annotations"
 )
 
-// IsDeclarativeFriendlyMessage returns true if the descriptor is
-// declarative-friendly.
+// DeclarativeFriendlyResource returns the declarative-friendly resource
+// associated with this descriptor.
 //
-// This is true if:
-// - The message is annotated with google.api.resource and
-//   style: DECLARATIVE_FRIENDLY is set.
-// - The message is a standard method request message for a resource with
-//   google.api.resource and style:DECLARATIVE_FRIENDLY set.
-func IsDeclarativeFriendlyMessage(m *desc.MessageDescriptor) bool {
-	// Get the google.api.resource annotation and see if it is styled
-	// declarative-friendly.
-	if resource := GetResource(m); resource != nil {
-		for _, style := range resource.GetStyle() {
-			if style == apb.ResourceDescriptor_DECLARATIVE_FRIENDLY {
-				return true
+// For messages:
+// If the message is annotated with google.api.resource and
+// style: DECLARATIVE_FRIENDLY is set, that message is returned.
+// If the message is a standard method request message for a resource with
+// google.api.resource and style:DECLARATIVE_FRIENDLY set, then the resource
+// is returned.
+//
+// For methods:
+// If the output message is a declarative-friendly resource, it is returned.
+// If the method begins with "List" and the first repeated field is a
+// declarative-friendly resource, the resource is returned.
+// If the method begins with "Delete", the return type is Empty, and an
+// appropriate resource message is found and is declarative-friendly, that
+// resource is returned.
+// If the method is a custom method where a matching resource is found (by
+// subset checks on the name) and is declarative-friendly, the resource is
+// returned.
+//
+// If there is no declarative-friendly resource, it returns nil.
+func DeclarativeFriendlyResource(d desc.Descriptor) *desc.MessageDescriptor {
```

Comment:
This is another refactor because I found I need _the resource itself_, not just a yes/no on declarative or not declarative, for one of the AIP-154 rules. (A forthcoming AIP-163 rule will have a similar need.)